include/envoy/buffer/buffer.h

```diff
@@ -32,6 +38,18 @@ class Instance {
    */
   virtual void add(const void* data, uint64_t size) PURE;
 
+  /**
+   * Add externally owned data to the buffer. Nothing is copied.
+   * The memory at <data> must be valid until all data has been read.
+   * <releaseCallback> is called when the memory is no longer needed by the buffer.
+   * @param data supplies a pointer to the externally owned data.
+   * @param releaseCallback a function to be called
+   * @param releaseArg optional argument passed to releaseCallback as <arg>.
+   * @param size supplies the size of the data.
+   */
+  virtual void addReference(
+    const void* data, uint64_t size, refReleaseCb releaseCallback, void* releaseArg) PURE;
```

Comment:
This interface would be much nicer if you either pass a std::function for the releaseCallback (with no releaseArg), or pass an interface that contains the data, size, and a destructor.

I realize you're aiming for high-performance here, but I'd rather keep the interface clean unless you have benchmarks proving that either of those alternatives is too slow.

Reply:
I'll look into how to clean this up. Is the general idea to have <= 2 args rather than like 4?